src/cargo/util/diagnostic_server.rs

```diff
@@ -258,14 +258,23 @@ impl RustfixDiagnosticServer {
 
     fn run(self, on_message: &dyn Fn(Message), done: &AtomicBool) {
         while let Ok((client, _)) = self.listener.accept() {
-            let client = BufReader::new(client);
-            match serde_json::from_reader(client) {
-                Ok(message) => on_message(message),
-                Err(e) => warn!("invalid diagnostics message: {}", e),
-            }
             if done.load(Ordering::SeqCst) {
                 break;
             }
+            let mut client = BufReader::new(client);
```

Comment:
FWIW with `read_to_string` the `BufReader` here is likely no longer necessary

Reply:
I looked at that for a bit.  It didn't seem too important, but if I understand the code correctly, a bare read_to_string uses a [32-byte buffer](https://github.com/rust-lang/rust/blob/2442823ef572a65092fbc46f6975633f983b50b6/src/libstd/io/mod.rs#L354) whereas BufReader uses an [8k buffer](https://github.com/rust-lang/rust/blob/2442823ef572a65092fbc46f6975633f983b50b6/src/libstd/io/buffered.rs#L71).  Does that sound correct?  Sometimes it's difficult to trace with various different traits involved.  `read_to_end_with_reservation` doesn't appear to grow its read size, which is a little surprising.

Reply:
Eh it could go either way I think. `read_to_string` starts with 32 bytes and exponentially increases (IIRC) and also avoids zeroing and does other fancy tricks. Ideally we'd reuse a buffer here but this isn't really performance sensitive anyway so it doesn't matter too much. Whichever you feel like doing is fine by me!

Reply:
Oh, indeed, it does 32,32,64,128,256,... (I'm guessing it's [this](https://github.com/rust-lang/rust/blob/2442823ef572a65092fbc46f6975633f983b50b6/src/liballoc/raw_vec.rs#L428)).  I'm fine with leaving it as-is, saves a few round trips in the common case (large text messages).

Reply:
Ah it's actually [this one](https://github.com/rust-lang/rust/blob/2442823ef572a65092fbc46f6975633f983b50b6/src/libstd/io/mod.rs#L344-L389), but it's morally the same thing!

Reply:
Oh, I was just looking at the end of that call chain, because `read_to_end_with_reservation` does not do any obvious doubling (and the behavior does not appear to be documented anywhere).  I'm assuming the call `g.buf.reserve(32)` goes to `RawVec::reserve` which goes to `reserve_internal` with the `Amortized` strategy to `amortized_new_size` which does the doubling (and the `needed_extra_cap` of 32 gets ignored pretty quickly).

Reply:
er oops, right!